Congratualtions! Preprocessing is done. Let's dive into the bag-of-words!

# Representación Bag-of-Words (Bolsa de Palabras)

La idea detrás de **Bag-of-Words (BoW)**, como sugiere su nombre, es bastante intuitiva: tomamos un documento y lo "lanzamos" en una bolsa. Al hacerlo, ignoramos la posición relativa entre las palabras. Es decir, lo que queda "en la bolsa" es esencialmente un **conjunto desordenado de palabras** [(Jurafsky & Martin, 2024)](https://web.stanford.edu/~jurafsky/slp3/ed3book.pdf).

A cambio, obtenemos:
- Una lista de palabras únicas.
- La frecuencia de cada palabra.

Por ejemplo, como se muestra en la ilustración a continuación, la palabra "coffee" aparece dos veces. 

![Ilustración Bag-of-Words](../images/bow-illustration-1.png)

Con una representación Bag-of-Words, utilizamos con fuerza la **frecuencia de palabras**, pero no tanto el **orden de las palabras**.

### Bag-of-Words en el Análisis de Sentimientos
En el contexto del análisis de sentimientos, el sentimiento de un texto (por ejemplo, un tweet) suele transmitirse más fuertemente por palabras específicas. Por ejemplo:
- Si un tweet contiene la palabra **"happy"**, probablemente transmita un sentimiento positivo.
- Sin embargo, no siempre es así. Por ejemplo, **"not happy"** expresa lo opuesto.

Cuando estas palabras específicas aparecen con mayor frecuencia, probablemente tendrán un impacto más fuerte en el análisis del sentimiento.


## Document Term Matrix (Matriz de Términos por Documento)

Ahora implementemos la idea de **Bag-of-Words**. Antes de profundizar, tomemos un momento para contextualizar. En la práctica, el análisis de texto generalmente implica trabajar con múltiples documentos. A partir de ahora, utilizaremos el término **documento** para referirnos a cualquier pieza de texto en la que realicemos análisis. Puede ser una frase, una oración, un tweet o cualquier otro texto—siempre que pueda representarse como una cadena de texto, la longitud no importa.

### Concepto
Imaginemos que tenemos cuatro documentos (es decir, las cuatro frases mostradas arriba) y los colocamos todos en la "bolsa". En lugar de obtener una lista de frecuencias de palabras, obtendremos una **matriz de términos por documento (DTM, por sus siglas en inglés)**.

En una DTM:
- La lista de palabras únicas (el **vocabulario**, \( V \)) contiene todas las palabras que aparecen en todos los documentos.
- Para cada **documento** (\( D \)), contamos cuántas veces aparece cada palabra del vocabulario en el documento y colocamos estos valores en la matriz.

En otras palabras, la DTM que construiremos es una matriz de \( D \times V \), donde:
- Cada **fila** corresponde a un documento.
- Cada **columna** corresponde a un token (o término único).

### Ejemplo
En el siguiente ejemplo, los tokens únicos de este conjunto de documentos se ordenan alfabéticamente para formar las columnas. Para cada documento, marcamos la ocurrencia de cada palabra presente. La representación numérica de cada documento es una fila en la matriz.

Por ejemplo, el primer documento, *"the coffee roaster"*, tiene la representación numérica:  
\[
[0, 1, 0, 0, 0, 1, 1, 0]
\]

A continuación, se muestra la matriz completa:

$$
\begin{array}{c|cccccccccccc}
 & \text{americano} & \text{coffee} & \text{iced} & \text{light} & \text{roast} & \text{roaster} & \text{the} & \text{time} \\\hline
\text{the coffee roaster} &0 &1	&0	&0	&0	&1	&1	&0 \\ 
\text{light roast} &0 &0	&0	&1	&1	&0	&0	&0 \\ 
\text{iced americano} &1 &0	&1	&0	&0	&0	&0	&0 \\ 
\text{coffee time} &0 &1	&0	&0	&0	&0	&0	&1 \\ 
\end{array}
$$

> **Nota**: Aunque la primera columna de la matriz muestra los documentos como texto, típicamente se les asignaría un índice numérico.

### Construcción de una DTM
Para crear una matriz de términos por documento, utilizaremos la clase `CountVectorizer` del paquete `sklearn`.


In [13]:
from sklearn.feature_extraction.text import CountVectorizer

La siguiente ilustración muestra el flujo de trabajo de tres pasos para crear una matriz de términos por documento (DTM) usando `CountVectorizer`.

<img src='../images/CountVectorizer1.png' alt="CountVectorizer" width="500">

### Flujo de Trabajo
1. **Tokenización**: Divide los documentos en palabras o tokens únicos.
2. **Construcción del Vocabulario**: Identifica todas las palabras únicas presentes en los documentos y las organiza en un vocabulario.
3. **Construcción de la DTM**: Calcula la frecuencia de cada palabra en cada documento y organiza estos valores en una matriz.

### Ejemplo Práctico
A continuación, vamos a recorrer estos pasos utilizando el ejemplo mencionado previamente.


### Ejemplo sencillo

In [14]:
# A toy example containing four documents
test = ['the coffee roaster',
        'light roast',
        'iced americano',
        'coffee time']

El primer paso es inicializar un objeto `CountVectorizer`. Dentro de los paréntesis redondos, podemos especificar configuraciones de parámetros si lo deseamos. Para explorar las opciones disponibles, podemos consultar la [documentación oficial](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html).

Por ahora, lo dejaremos en blanco para usar las configuraciones predeterminadas.


In [15]:
# Create a CountVectorizer object
vectorizer = CountVectorizer()

El segundo paso es usar el método `fit` del objeto `CountVectorizer` en los datos, lo que implica crear un vocabulario de tokens a partir del conjunto de documentos. En el tercer paso, utilizamos el método `transform` para transformar los datos de acuerdo con el objeto `CountVectorizer` "ajustado". Esto significa tomar cada documento y contar las ocurrencias de los tokens según el vocabulario establecido durante el paso de "ajuste" (`fit`).

Aunque suena algo complejo, los pasos 2 y 3 se pueden realizar de manera conjunta usando la función `fit_transform`.


In [16]:
# Fit and transform to create a DTM
test_count = vectorizer.fit_transform(test)

El resultado de `fit_transform` debería ser la matriz de términos del documento (DTM, por sus siglas en inglés).

¡Echemos un vistazo!


In [17]:
test_count

<4x8 sparse matrix of type '<class 'numpy.int64'>'
	with 9 stored elements in Compressed Sparse Row format>

Aparentemente tenemos una "matriz dispersa"—una matriz que contiene muchos ceros. Esto tiene sentido. Para cada documento, hay palabras que no aparecen en absoluto, y estas se cuentan como ceros en la DTM. Esta matriz dispersa se almacena en un formato "Compressed Sparse Row" (CSR), un formato diseñado para ahorrar memoria al manejar matrices dispersas.

Convirtámosla en una matriz densa, donde probablemente se representen esos ceros, como en un array de numpy.


In [18]:
# Convert DTM to a dense matrix 
test_count.todense()

matrix([[0, 1, 0, 0, 0, 1, 1, 0],
        [0, 0, 0, 1, 1, 0, 0, 0],
        [1, 0, 1, 0, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0, 0, 1]])

¡Así que esta es nuestra DTM! La matriz es la misma que se muestra arriba. Para que sea más fácil de leer, vamos a convertirla en un dataframe. Los nombres de las columnas deben ser los tokens del vocabulario, los cuales podemos acceder con la función `get_feature_names_out`.


In [19]:
# Retrieve the vocabulary
vectorizer.get_feature_names_out()

array(['americano', 'coffee', 'iced', 'light', 'roast', 'roaster', 'the',
       'time'], dtype=object)

In [20]:
# Create a DTM dataframe
test_dtm = pd.DataFrame(data=test_count.todense(),
                        columns=vectorizer.get_feature_names_out())

¡Aquí está! La DTM de nuestros datos de ejemplo ahora es un dataframe. El índice de `test_dtm` corresponde a la posición de cada documento en la lista `test`.


In [21]:
test_dtm

,americano,coffee,iced,light,roast,roaster,the,time
0,0,1,0,0,0,1,1,0
1,0,0,0,1,1,0,0,0
2,1,0,1,0,0,0,0,0
3,0,1,0,0,0,0,0,1


Esperemos que este ejemplo sencillo proporcione una explicación clara sobre cómo crear una Matriz de Términos por Documentos (DTM).

## DTM para Tweets

Comenzaremos inicializando un objeto `CountVectorizer`. En la siguiente celda, hemos incluido algunos parámetros que las personas suelen ajustar. Estos parámetros están actualmente configurados con sus valores predeterminados.

Cuando construimos una DTM, el comportamiento predeterminado es convertir el texto de entrada a minúsculas. Si no se proporciona nada para `stop_words`, el valor predeterminado es conservarlas. Los siguientes tres parámetros se utilizan para controlar el tamaño del vocabulario, a los cuales regresaremos en un momento.


In [22]:
# Create a CountVectorizer object
vectorizer = CountVectorizer(lowercase=True,
                             stop_words=None,
                             min_df=1,
                             max_df=1.0, 
                             max_features=None)

In [23]:
# Fit and transform to create DTM
counts = vectorizer.fit_transform(tweets['text_processed'])
counts

<11541x8751 sparse matrix of type '<class 'numpy.int64'>'
	with 191139 stored elements in Compressed Sparse Row format>

In [24]:
# Do not run if you have limited memory - this includes DataHub and Binder
np.array(counts.todense())

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [25]:
# Extract tokens
tokens = vectorizer.get_feature_names_out()

In [26]:
# Create DTM
first_dtm = pd.DataFrame(data=counts.todense(),
                         index=tweets.index,
                         columns=tokens)

# Print the shape of DTM
print(first_dtm.shape)

(11541, 8751)


Si dejamos el `CountVectorizer` con la configuración predeterminada, el tamaño del vocabulario de los datos de los tweets será de 8751.


In [27]:
first_dtm.head()

,_exact_,_wtvd,aa,aaaand,aadv,aadvantage,aal,aaron,ab,aback,...,zero,zig,zip,zippers,zone,zones,zoom,zukes,zurich,zz
0,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


La mayoría de los tokens tienen cero ocurrencias, al menos en los primeros cinco tweets. 

¡Echemos un vistazo más de cerca al DTM!


In [28]:
# Most frequent tokens
first_dtm.sum().sort_values(ascending=False).head(10)

user      12882
to         6987
digit      6927
the        5088
you        3635
for        3386
flight     3320
and        3276
on         3142
my         2751
dtype: int64

In [29]:
# Least frequent tokens
first_dtm.sum().sort_values(ascending=True).head(10)

_exact_                     1
mightmismybrosgraduation    1
midterm                     1
midnite                     1
midland                     1
michelle                    1
michele                     1
michael                     1
mhtt                        1
mgmt                        1
dtype: int64

No es sorprendente ver que "user" y "digit" están entre los tokens más frecuentes, ya que reemplazamos cada identificador idiosincrático con estos marcadores de posición. El resto de los tokens más frecuentes son principalmente palabras vacías (stop words).

Quizás un patrón más interesante sea buscar cuál token aparece más en cualquier tweet dado:


In [30]:
counts = pd.DataFrame()

# Retrieve the index of the tweet where a token appears most frequently
counts['token'] = first_dtm.idxmax(axis=1)

# Retrieve the number of occurrence 
counts['number'] = first_dtm.max(axis=1)

# Filter out placeholders
counts[(counts['token']!='digit')
       & (counts['token']!='hashtag')
       & (counts['token']!='user')].sort_values('number', ascending=False).head(10)

,token,number
3127,lt,6
918,worst,6
10572,to,5
8148,the,5
10742,to,5
152,to,5
5005,to,5
10923,the,5
7750,to,5
355,to,5


Parece que, entre todos los tweets, como máximo un token aparece seis veces, y este token es la palabra "It" o la palabra "worst."

Volvamos a nuestro dataframe de tweets y localicemos el tweet número 918.


In [31]:
# Retrieve 918th tweet: "worst"
tweets.iloc[918]['text']

"@united is the worst. Worst reservation policies. Worst costumer service. Worst worst worst. Congrats, @Delta you're not that bad!"

## Personalizar el `CountVectorizer`

Hasta ahora, siempre hemos utilizado la configuración predeterminada de parámetros para crear nuestras DTMs, pero en muchos casos podríamos querer personalizar el objeto `CountVectorizer`. El propósito de hacerlo es filtrar aún más los tokens innecesarios. En el ejemplo siguiente, ajustamos los siguientes parámetros:

- `stop_words = 'english'`: ignorar las palabras vacías en inglés.
- `min_df = 2`: ignorar las palabras que no aparecen al menos dos veces.
- `max_df = 0.95`: ignorar las palabras si aparecen en más del 95% de los documentos.

🔔 **Pregunta**: ¡Pausémonos un momento para discutir si tiene sentido configurar estos parámetros! ¿Qué opinas?

A menudo, no nos interesan las palabras cuyas frecuencias son demasiado bajas o demasiado altas, por lo que usamos `min_df` y `max_df` para filtrarlas. Alternativamente, podemos definir el tamaño de nuestro vocabulario como $N$ configurando `max_features`. En otras palabras, le indicamos a `CountVectorizer` que solo considere los $N$ tokens más frecuentes al construir la DTM.


In [32]:
# Customize the parameter setting
vectorizer = CountVectorizer(lowercase=True,
                             stop_words='english',
                             min_df=2,
                             max_df=0.95,
                             max_features=None)

In [33]:
# Fit, transform, and get tokens
counts = vectorizer.fit_transform(tweets['text_processed'])
tokens = vectorizer.get_feature_names_out()

# Create the second DTM
second_dtm = pd.DataFrame(data=counts.todense(),
                          index=tweets.index,
                          columns=tokens)

Nuestra segunda DTM tiene un vocabulario sustancialmente más pequeño en comparación con la primera.


In [34]:
print(first_dtm.shape)
print(second_dtm.shape)

(11541, 8751)
(11541, 4471)


In [35]:
second_dtm.head()

,aa,aadv,aadvantage,aal,abandoned,abc,ability,able,aboard,abq,...,yummy,yup,yvonne,yvr,yyj,yyz,zero,zone,zoom,zurich
0,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


Nuestra segunda DTM tiene un vocabulario sustancialmente más pequeño en comparación con la primera.

La lista de tokens más frecuentes ahora incluye palabras que tienen más sentido para nosotros, como "cancelled" y "service."


In [36]:
second_dtm.sum().sort_values(ascending=False).head(10)

digit        6927
flight       3320
hashtag      2633
cancelled     956
thanks        921
service       910
just          801
customer      726
time          695
help          687
dtype: int64

## 🥊 Reto 2: Lematizar el Texto de Entrada

Recordemos de la Parte 1 que introdujimos el uso de `spaCy` para realizar lematización, es decir, "recuperar" la forma base de una palabra. Este proceso reduce el tamaño del vocabulario al mantener mínimas las variaciones de las palabras, lo que puede ayudar a mejorar el rendimiento del modelo en la clasificación de sentimientos.

Ahora vamos a implementar la lematización en nuestros datos de tweets y utilizaremos el texto lematizado para crear una tercera DTM.

Completa la función `lemmatize_text`. Esta función requiere un texto de entrada y devuelve los lemas de todos los tokens.

Aquí tienes algunas pistas para guiarte a través de este desafío:

- **Paso 1**: inicializa una lista para almacenar los lemas.
- **Paso 2**: aplica el pipeline `nlp` al texto de entrada.
- **Paso 3**: itera sobre los tokens en el texto procesado y recupera el lema de cada token.  
  - **PISTA**: la lematización es una de las anotaciones lingüísticas que el pipeline `nlp` realiza automáticamente. Podemos usar `token.lemma_` para acceder a la anotación.


In [37]:
# Import spaCy
import spacy
nlp = spacy.load('en_core_web_sm')

In [ ]:
# Create a function to lemmatize text
def lemmatize_text(text):
    '''Lemmatize the text input with spaCy annotations.'''

    # Step 1: Initialize an empty list to hold lemmas
    lemma = ...

    # Step 2: Apply the nlp pipeline to input text
    doc = ...

    # Step 3: Iterate over tokens in the text to get the token lemma
    for token in doc:
        lemma.append(...)

    # Step 4: Join lemmas together into a single string
    text_lemma = ' '.join(lemma)
    
    return text_lemma

Apliquemos la función primero al siguiente tweet de ejemplo

In [39]:
# Apply the function to an example tweet
print(tweets.iloc[33]["text_processed"])
print(f"{'='*50}")
print(lemmatize_text(tweets.iloc[33]['text_processed']))

USER wow this just blew my mind
USER wow this just blow my mind


Y luego, lematizaremos los datos de los tweets y guardaremos el resultado en una nueva columna `text_lemmatized`.


In [40]:
# This may take a while!
tweets['text_lemmatized'] = tweets['text_processed'].apply(lambda x: lemmatize_text(x))

Ahora, con la columna `text_lemmatized`, creemos un tercer DTM. La configuración de parámetros es la misma que la del segundo DTM.


In [41]:
# Create the vectorizer (the same param setting as previous)
vectorizer = CountVectorizer(lowercase=True,
                             stop_words='english',
                             min_df=2,
                             max_df=0.95,
                             max_features=None)

# Fit, transform, and get tokens
counts = vectorizer.fit_transform(tweets['text_lemmatized'])
tokens = vectorizer.get_feature_names_out()

# Create the third DTM
third_dtm = pd.DataFrame(data=counts.todense(),
                         index=tweets.index,
                         columns=tokens)
third_dtm.head()

,aa,aadv,aadvantage,aal,abandon,abc,ability,able,aboard,abq,...,yummy,yup,yvonne,yvr,yyj,yyz,zero,zone,zoom,zurich
0,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [42]:
# Print the shapes of three DTMs
print(first_dtm.shape)
print(second_dtm.shape)
print(third_dtm.shape)

(11541, 8751)
(11541, 4471)
(11541, 3553)


Imprimamos los 10 tokens más frecuentes como de costumbre. Estos tokens ahora son lemas y sus conteos también cambian después de la lematización.


In [43]:
# Get the most frequent tokens in the third DTM
third_dtm.sum().sort_values(ascending=False).head(10)

digit       6927
flight      4043
hashtag     2633
thank       1455
hour        1134
cancel       948
delay        937
service      937
customer     902
time         856
dtype: int64

In [44]:
# Compared to the most frequent tokens in the second DTM
second_dtm.sum().sort_values(ascending=False).head(10)

digit        6927
flight       3320
hashtag      2633
cancelled     956
thanks        921
service       910
just          801
customer      726
time          695
help          687
dtype: int64